utils/include/utils/text_matchers.h

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (c) 2022 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+
+#ifndef HLASMPLUGIN_UTILS_TEXT_MATCHERS_H
+#define HLASMPLUGIN_UTILS_TEXT_MATCHERS_H
+
+#include <cctype>
+#include <optional>
+#include <string_view>
+#include <type_traits>
+#include <utility>
+
+namespace hlasm_plugin::utils::text_matchers {
+
+template<typename It>
+[[maybe_unused]] std::true_type same_line_detector(const It& t, decltype(t.same_line(t)) = false);
+[[maybe_unused]] std::false_type same_line_detector(...);
+
+template<typename It>
+constexpr bool same_line(const It& l, const It& r)
+{
+    if constexpr (decltype(same_line_detector(l))::value)
+        return l.same_line(r);
+    else
+        return true;
+}
+
+template<bool case_sensitive, bool single_line>
+class basic_string_matcher
+{
+    std::string_view to_match;
+
+public:
+    // case_sensitive => to_match assumed in capitals
+    explicit constexpr basic_string_matcher(std::string_view to_match)
+        : to_match(to_match)
+    {}
+
+    template<typename It>
+    bool operator()(It& b, const It& e) const noexcept
+    {
+        It work = b;
+
+        for (auto c : to_match)
+        {
+            if (work == e)
+                return false;
+            if constexpr (single_line)
+            {
+                if (!same_line(work, b))
+                    return false;
+            }
+            auto in_c = *work++;
+            if constexpr (!case_sensitive)
+                in_c = std::toupper((unsigned char)in_c);
+
+            if (in_c != c)
+                return false;
+        }
+        b = work;
+        return true;
+    }
+};
+
+template<bool negate>
+class char_matcher_impl
+{
+    std::string_view to_match;
+
+public:
+    // case_sensitive => to_match assumed in capitals
+    explicit constexpr char_matcher_impl(std::string_view to_match)
+        : to_match(to_match)
+    {}
+
+    template<typename It>
+    constexpr bool operator()(It& b, const It& e) const noexcept
+    {
+        if (b == e)
+            return false;
+        bool result = (to_match.find(*b) != std::string_view::npos) != negate;
+        if (result)
+            ++b;
+        return result;
+    }
+};
+
+using char_matcher = char_matcher_impl<false>;
+using not_char_matcher = char_matcher_impl<true>;
+
+template<typename Matcher>
+constexpr auto star(Matcher&& matcher)
```

Comment:
The naming certainly does require more thinking...
In this case, it is just direct reference to the regex languages... even though this one does not backtrack and retry.
Therefore I am not sure about try_many, because there is no trying done, it is more like as_much_as_you_can... (but that's too long)

Reply:
What about `hungry` or `greedy`?

Reply:
still need to distinguish star and plus... both are greedy...